celeschai iP

celeschai opened the pull request:
# Bee #
> [!TIP] 
> “Your mind is for having ideas, _not_ holding them.” – David Allen (source)

### Bee frees your mind of having to remember things you need to do. ###
- text-based
- easy to learn
- SUPER FAST to use

### Features: ###

- [x] Managing tasks
- [x] Managing deadlines 
- [x] Managing events

## All you need to do is, ##

1. download the `.jar` file from [here](https://github.com/celeschai/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is **FREE**!

### For developers ###
to start the application through CLI, download the codebase and run `./gradlew run`.
CLI version for UI is available before `branch-Level-10` was merged.

Entrypoint to GUI application:
```java
public class Main extends Application {
    private Bee bee = new Bee();

    @Override
    public void start(Stage stage) {
        try {
            //...
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

celeschai committed:
feat: Add chatbot

celeschai committed:
feat: Respond to user input

celeschai committed:
feat: Add task and print list

celeschai committed:
feat: Update task status

celeschai committed:
feat: Add Todo, Deadline, Event

User can add todo (just task name),
deadline (name + due date),
event (name + start + end time).

Todo, Deadline, and Event have common behaviours
such as markAsDone() and markAsIncomplete().

The common behaviours cause code duplication.

Extracting the common behaviour into a super class
Task allows us to use polymorphism when manipulating
the tasks in a single todolist.

Let's modify the parent class Task to be abstract
so that it will not be instantiated.

Let's extend the parent class task to implement
different types of task to utilise inheritance.

Update to common behaviours will be done in a
single place which is the parent class Task.

celeschai committed:
feat: Handle invalid input

Use exceptions to display error message for users.

Use try-catch block places error handling logic in
a single code block to reduce reduplication of code.

Let's pull up exception classes related to invalid
task inputs into a new parent class named TaskExceptions
to standardise error message and instructions given to users.

Let's pull up all exception classes uses in Bee
into a new parent class name BeeExceptions to
maintain consistency of wording by the chatbot.

celeschai committed:
refactor: Handle invalid index using exception

celeschai committed:
feat: Delete task

celeschai committed:
test: Semi-automate testing with I/O redirection

celeschai committed:
fix: Handle invalid task indices that are not int
feat: Provide user guid with 'help'

celeschai committed:
refactor: Introduce more OOP with new classes

celeschai committed:
feat: Add save to local drive capability'

celeschai committed:
feat: Add save to local drive capabilities

celeschai committed:
feat: Represent date time in java objects

celeschai committed:
Merge branch 'add-gradle-support'
Automate project builds using Gradle

celeschai committed:
test: Test Parser class with Junit

celeschai committed:
feat: Find tasks by name

celeschai committed:
Merge branch 'branch-Level-9'

celeschai committed:
docs: Handle text file for date time testing

celeschai committed:
Merge branch 'branch-Level-8'

celeschai committed:
docs: Add JavaDoc comments to code

celeschai committed:
Merge branch 'branch-A-JavaDoc'

celeschai committed:
docs: Tweak code & comments to comply with coding standard

celeschai committed:
Merge branch 'branch-A-CodingStandard'

celeschai committed:
chore: Update gitignore for JAR files

SQ77 commented on src/main/java/chatbot/Bee.java:
```diff
@@ -0,0 +1,194 @@
+                    String next = sc.nextLine();
+
+                    if (next.matches("bye.*")) {
+                        boolean saved = storage.writeToTaskListFile(taskList.parseTaskListToTxt());
```
Should this boolean be named "isSaved" instead?

SQ77 commented on src/main/java/chatbot/Ui.java:
```diff
@@ -0,0 +1,95 @@
+    /**
+     * Prints out user guide.
+     */
+    public static final String help = """
```
Should this be named as "HELP" instead since it is a constant?

SQ77 replied:
I like how you included a command to print out the user guide 👍

SQ77 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,47 @@
+ */
+public abstract class Task {
+    private String name = "new task";
+    private boolean status = false;
```
Should this variable be named as "isDone" to be more descriptive?

SQ77 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    /**
+     * Adds Task to TaskList.
+     * @param task Task object to be added to the ArrayLst.
```
Should there be a blank line between these two lines?

SQ77 reviewed (commented):
Overall, your code was easy to read. I liked your use of regular expressions when handling user commands. Just a few very minor coding standard violations to fix.

jasmiinee commented on src/main/java/chatbot/Bee.java:
```diff
@@ -0,0 +1,194 @@
+import java.time.format.DateTimeParseException;
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
```
Should these java imports be at the top of the imports grouping instead?

jasmiinee commented on src/main/java/chatbot/Bee.java:
```diff
@@ -0,0 +1,194 @@
+package chatbot;
+
+import task.*;
```
Should the imports be explicitly listed out instead of using the '*' operator ?

jasmiinee commented on src/main/java/chatbot/Bee.java:
```diff
@@ -0,0 +1,194 @@
+        Storage storage = new Storage();
+
+        // Checks if source text file is successfully opened
+        boolean isOpenFile = storage.readFromTaskListFile();
```
Would the boolean "isFileOpen" be better?

jasmiinee commented on src/main/java/chatbot/Bee.java:
```diff
@@ -0,0 +1,194 @@
+
+                } catch (DateTimeParseException e) {
+                    // Guides user on date time input formats
+                    Ui.printBtnLines("Format your time in: " +
```
Should the line break be before the '+' operator instead?

jasmiinee commented on src/main/java/chatbot/Parser.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.time.format.ResolverStyle;
```
Should these java imports be at the top of all the imports instead? and should the imports also be in lexicographic order instead?

jasmiinee commented on src/main/java/chatbot/BeeException.java:
```diff
@@ -0,0 +1,20 @@
+     * @param msg Error message to be shown.
+     */
+    public BeeException(String msg) {
+        super("Hey!!! " + msg + "\nType 'help' to see what you can do.");
```
I like how you included a help command

jasmiinee reviewed (commented):
Overall, code is neat and rather easy to read except for a few deeply nested components and violations of coding standards.

celeschai committed:
fix: Updates according to comments

celeschai committed:
style: Check style and fix

celeschai committed:
feat: Add GUI

celeschai committed:
docs: Update README

celeschai committed:
Merge branch 'branch-A-CheckStyle'

celeschai committed:
Merge branch 'branch-Level-10'

celeschai committed:
build: Update build.gradle and locally saved file

celeschai committed:
test: Assert expected states of programme

celeschai committed:
refactor: Extract case handling functions

Parser.parseFromTxtTaskList and Bee.getBeeResponse
handles many different cases of user input.

The huge number of possible cases creates lengthy code.

Extracting the handliing of each specific case into a
method allows us to keep the code clean, concise, and
adhere to SLAP more closely.

Let's pull up handling of each input case into
private methods for that particular class.

Private methods are preferrable as these methods
are specific to the parser and main chatbot responses
which means that they should not be used or
accessible from other classes.

celeschai committed:
Merge branch 'branch-A-Assertions'

celeschai committed:
Merge branch 'master' into branch-A-CodeQuality

celeschai committed:
Merge branch 'branch-A-CodeQuality'

celeschai committed:
fix: Change parameters of extracted function

Matcher groups separates the string according to the
regex pattern specified.

During refactoring and extraction of function it is crucial to
ensure the correct group was passed into the handling function.

Passing in the correct group of matched strings allows
the extracted function to perform the expected behaviour as if
the code was written sequentially instead of reprocessing the
entire string.

Let's check that Matcher groups are properly passed into the
extracted functions especially in Parser and Bee where
extracted functions are heavily utilized.

celeschai committed:
build: Set up GitHub Actions

celeschai committed:
feat: Check start time is before end time for events

celeschai committed:
feat: Search tasks by partial name

celeschai committed:
build: Update GitHub Actions workflow

celeschai committed:
docs: Update user guide

celeschai committed:
Merge pull request #2 from celeschai/branch-A-UserGuide

docs: Update user guide

celeschai committed:
style: Format code style

celeschai committed:
build: Update gradle config

celeschai committed:
style: Update code formaat

celeschai committed:
test: Update test cases

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide screenshot

celeschai committed:
fix: Add flexibility to list and help command

celeschai committed:
docs: Reformat representative screenshot

celeschai committed:
docs: Fix typo

celeschai committed:
docs: Set window title

celeschai committed:
docs: Reformat representative screenshot

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide

celeschai committed:
docs: Update user guide
